src/shells/shared/content-script.ts

```diff
@@ -52,17 +60,47 @@
 		);
 	}
 
-	pingBackend();
-
-	// On page reload the content script may be initialized before the backend
-	if (!backendInitialized) {
-		let interval: any;
-		interval = setInterval(() => {
-			if (backendInitialized) {
-				clearInterval(interval);
-			} else {
-				pingBackend();
+	function forwardWindowPostMessagesToPortMessage() {
+		function onWindowPostMessage(ev: MessageEvent) {
+			if (ev.source === window && ev.data) {
+				// TODO(Sven): Why are we forwarding this message from port.onMessage to window.postMessage
+				// when we are listening in the same closure?
```

Comment:
The `content-script` has access to the page, but doesn't run _in_ the same scope. To be able to post messages from the client adapter we can only use `window.postMessage`. The basic data flow is this:

```
                                   +------------------+
                                   | Panel (icon bar) |
                                   +--------+---------+
                                            |
+------+    +----------------+    +---------+---------+
| Page +--->+ content-script +--->+ background-script |
+------+    +----------------+    +---------+---------+
                                            |
                                    +-------+--------+
                                    | devtools panel |
                                    +----------------+

```

And it works the other way round too. The devtools panel needs to notify the page when it does highlighting or even just for requesting prop data for the sidebar. The distinction for what goes where is done via `event.data.source`. Note that the event that is sent has a different shape, depending wether it was sent via `window.postMessage` or `port.postMesssage`. That tripped me up a few times.

Reply:
I added that diagram and some other things to `INTERNALS.md`. I got one question there for you regarding the split of `content-script.ts` and `initClient.ts`. I guess the reason is because `initClient.ts` is injected once and `content-script.ts` might be injected multiple times. But would be nice if you could clarify this.

Also we should add some sentences about message routing (i.e. the `event.data.source` you mentioned here with a bit more detail).